### PR TITLE
Adding a simple VAE Example

### PR DESCRIPTION
**Pull request by neerajprad**
Addresses #29.

TODOs:
 - [x] binarize MNIST images instead of using the hack of disabling args validation and relying on `Bernoulli.logpmf`'s BCE implementation. 
 - [x] verify that the model is indeed learning something useful - ~the loss seems to remain roughly the same across epochs.~

Later PR:
 - [ ] Remove or figure out a better way to include the `args_check` flag, which is really hard to exercise with our current distribution class organization. Right now this is a global flag, but its really unintuitive since the flag might be off when `distribution.logpmf` method is called. If we have this flag, we should follow the pattern that we have in PyTorch currently. - #56.

### CODE REVIEW

**Comment by neerajprad** [comment]
This should be ready for a review now.

File: numpyro/examples/vae.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    @jit
+    def epoch_train(opt_state, rng):
+        def body_fn(i, val):
+            loss_sum, opt_state, rng = val
+            rng, batch = binarize(rng, train_fetch(i, train_idx)[0])
+            loss, opt_state, rng = svi_update(i, opt_state, rng, (batch,), (batch,),)
+            loss_sum += loss
+            return loss_sum, opt_state, rng
+
+        return lax.fori_loop(0, num_train, body_fn, (0., opt_state, rng))
```
Comment (neerajprad) [inline numpyro/examples/vae.py]: Even here jitting the for loop brings down the runtime to 25% of just jitting `body_fn`. Thanks for the tip, @fehiepsi.

**Comment by fehiepsi** [comment]
Nice! I'll go over this tomorrow.

File: numpyro/svi.py [hunk 2]
```diff
@@ -35,7 +39,12 @@ def update_fn(i, opt_state, rng, model_args=(), guide_args=()):
         rng, = random.split(rng, 1)
         return loss_val, opt_state, rng
 
-    return init_fn, update_fn
+    def evaluate(opt_state, rng, model_args=(), guide_args=()):
```
Comment (fehiepsi) [inline numpyro/svi.py]: +1

File: numpyro/examples/vae.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        img_loc = decode(params['decoder'], z).reshape([28, 28])
+        plt.imsave(os.path.join(RESULTS_DIR, 'recons_epoch={}.png'.format(epoch)), img_loc, cmap='gray')
+
+    for i in range(args.num_epochs):
+        t_start = time.time()
+        num_train, train_idx = train_init()
+        _, opt_state, rng = epoch_train(opt_state, rng)
+        rng, = random.split(rng, 1)
+        num_test, test_idx = test_init()
+        test_loss = eval_test(opt_state, rng)
```
Comment (fehiepsi) [inline numpyro/examples/vae.py]: nit: I think it is better to split rng into 2 (1 to train and update rng, 1 for eval) to avoid a confusion that: rng here is used for eval and the train of next epoch.

Comment (neerajprad) [reply numpyro/examples/vae.py]: Done.
